api/ocm/tools/signing/store.go

```diff
@@ -0,0 +1,211 @@
+package signing
+
+import (
+	"io"
+	"os"
+	"sync"
+
+	"github.com/mandelsoft/filepath/pkg/filepath"
+	"github.com/mandelsoft/goutils/errors"
+	"github.com/mandelsoft/goutils/sliceutils"
+	"github.com/mandelsoft/vfs/pkg/vfs"
+	"golang.org/x/exp/maps"
+
+	"ocm.software/ocm/api/ocm/compdesc"
+	metav1 "ocm.software/ocm/api/ocm/compdesc/meta/v1"
+	"ocm.software/ocm/api/utils"
+	common "ocm.software/ocm/api/utils/misc"
+	"ocm.software/ocm/api/utils/runtime"
+)
+
+type VerifiedStore interface {
+	Add(cd *compdesc.ComponentDescriptor, signatures ...string)
+	Remove(n common.VersionedElement)
+	Get(n common.VersionedElement) *compdesc.ComponentDescriptor
+	GetEntry(n common.VersionedElement) *StorageEntry
+
+	GetResourceDigest(n common.VersionedElement, id metav1.Identity) *metav1.DigestSpec
+	GetResourceDigestByIndex(n common.VersionedElement, idx int) *metav1.DigestSpec
+
+	Entries() []common.NameVersion
+
+	Load() error
+	Save() error
+}
+
+type verifiedStore struct {
+	lock    sync.Mutex
+	storage *StorageDescriptor
+	fs      vfs.FileSystem
+	file    string
+}
+
+var _ VerifiedStore = (*verifiedStore)(nil)
+
+func NewLocalVerifiedStore() VerifiedStore {
+	return &verifiedStore{storage: &StorageDescriptor{}}
+}
+
+func NewVerifiedStore(path string, fss ...vfs.FileSystem) (VerifiedStore, error) {
+	eff, err := utils.ResolvePath(path)
+	if err != nil {
+		return nil, err
+	}
+
+	fs := utils.FileSystem(fss...)
+
+	s := &verifiedStore{
+		fs:   fs,
+		file: eff,
+	}
+
+	err = s.Load()
+	if err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func (v *verifiedStore) Load() error {
+	v.lock.Lock()
+	defer v.lock.Unlock()
+	if v.fs == nil {
+		return nil
+	}
+
+	dir := filepath.Dir(v.file)
+
+	if ok, err := vfs.DirExists(v.fs, dir); !ok || err != nil {
+		if err != nil {
+			return err
+		}
+		return errors.ErrNotFound("directory", dir)
+	}
+
+	var storage StorageDescriptor
+	f, err := v.fs.Open(v.file)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return err
+		}
+	} else {
```

Comment:
Huh, why is it okay if it does not exist?

Reply:
The store must does not exist, when it is used for the first time. In this case only the folder must exists.
If it is nmodified later on, it saved on-the-fly, so that the next time it is used the correct content is available.

Otherwise the user always would have to create a file with the empty json/yaml representation. (potentially an empty file), before he could use the store option.

Reply:
Gotcha, thanks for the explanation.